server/server.go

```diff
@@ -257,15 +343,238 @@ func (p *proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 				w.Header().Add(http.TrailerPrefix+name, v)
 			}
 		}
+	}
+	return nil
+}
+
+func parseServerMessage(buf interface{}) ([]byte, websocket.MessageType, error) {
+	if data, ok := buf.(string); ok {
+		return []byte(data), websocket.MessageText, nil
+	}
+
+	if arrData, ok := buf.([]interface{}); ok {
+		if b64data, ok := arrData[0].(string); ok {
+			data, err := base64.StdEncoding.DecodeString(b64data)
+			if err == nil {
+				return data, websocket.MessageBinary, nil
+			}
+		}
+	}
+
+	return nil, websocket.MessageBinary, errors.New("unexpected data format from server")
+}
+
+func (p *proxy) handleWebsocketRequest(w http.ResponseWriter, r *http.Request) error {
+	ws := newWsSessionHelper()
+
+	// websocket: shimPath/open
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("http://:%v%v/open", port, shimPath), nil)
```

Comment:
I think the URL for these shim requests has to be based off of the incoming (websocket upgrade) request in order to maintain consistency and make it easier to properly configure the backend server.